agent/eventhandler/task_handler.go

```diff
@@ -15,244 +15,323 @@ package eventhandler
-					backoff.Reset()
-					taskEvents.events.Remove(eventToSubmit)
-				} else {
-					seelog.Errorf("TaskHandler, Unretriable error submitting container state change [%s]: %v",
-						event.String(), err)
-				}
-			} else if event.taskShouldBeSent() {
-				seelog.Infof("TaskHandler, Sending task change: %s", event.String())
-				err = client.SubmitTaskStateChange(event.taskChange)
-				if err == nil {
-					// submitted or can't be retried; ensure we don't retry it
-					event.setSent()
-					if event.taskChange.Task != nil {
-						event.taskChange.Task.SetSentStatus(event.taskChange.Status)
-					}
-					handler.stateSaver.Save()
-					seelog.Debugf("TaskHandler, Submitted task state change: %s", event.String())
-					backoff.Reset()
-					taskEvents.events.Remove(eventToSubmit)
-				} else {
-					seelog.Errorf("TaskHandler, Unretriable error submitting task state change[%s]: %v",
-						event.String(), err)
-				}
-			} else if event.taskAttachmentShouldBeSent() {
-				seelog.Infof("TaskHandler, Sending task attachment change: %s", event.String())
-				err = client.SubmitTaskStateChange(event.taskChange)
-				if err == nil {
-					// submitted or can't be retried; ensure we don't retry it
-					event.setSent()
-					if event.taskChange.Attachment != nil {
-						event.taskChange.Attachment.SetSentStatus()
-						event.taskChange.Attachment.StopAckTimer()
-					}
-					handler.stateSaver.Save()
-					seelog.Debugf("TaskHandler, Submitted task attachment state change: %s", event.String())
-					backoff.Reset()
-					taskEvents.events.Remove(eventToSubmit)
-				} else {
-					seelog.Errorf("TaskHandler, Unretriable error submitting task attachment state change [%s]: %v",
-						event.String(), err)
-				}
-			} else {
-				// Shouldn't be sent as either a task or container change event; must have been already sent
-				seelog.Infof("TaskHandler, Not submitting redundant event; just removing: %s", event.String())
-				taskEvents.events.Remove(eventToSubmit)
-			}
+// sendChange adds the change to the sendable events list. If the
+// event for the task is in the process of being sent, it doesn't trigger
+// the async submit state api method
+func (taskEvents *taskSendableEvents) sendChange(change *sendableEvent,
+	client api.ECSClient,
+	handler *TaskHandler) {
 
-			if taskEvents.events.Len() == 0 {
-				seelog.Debug("TaskHandler, Removed the last element, no longer sending")
-				taskEvents.sending = false
-				done = true
-				return nil
-			}
+	taskEvents.lock.Lock()
+	defer taskEvents.lock.Unlock()
 
-			return err
-		})
+	// Update taskEvent
+	seelog.Infof("TaskHandler: Adding event: %s", change.toString())
+	taskEvents.events.PushBack(change)
+
+	if !taskEvents.sending {
+		taskEvents.sending = true
+		go handler.submitTaskEvents(taskEvents, client, change.taskArn())
+	} else {
+		seelog.Debugf("TaskHandler: Not submitting change as the task is already being sent: %s", change.toString())
```

Comment:
So I think there's a race here, but the impact of the race is probably only a delay of `drainEventsFrequency`.  If we evaluate `taskEvents.sending` as true we won't cause a new goroutine to be created for `handler.submitTaskEvents`.  However, if the existing goroutine is just about done sending, it's possible it won't pick up the new event that we just added on line 263.  I think this is a reasonable tradeoff here, but it would be worth recording it somewhere that the maximum length of time we can have an unsubmitted change is really effectively `2*drainEventsFrequency`.